datafusion-cli/src/catalog.rs

```diff
@@ -115,13 +114,14 @@ impl CatalogProvider for DynamicFileCatalogProvider {
     }
 }
 
-/// Wraps another schema provider
-struct DynamicFileSchemaProvider {
+/// Wraps another schema provider. [DynamicObjectStoreSchemaProvider] is responsible for registering the required
```

Comment:
I am trying to understand why datafusion-cli can't simply use `DynamicFileCatalog` -- why does it need another layer of wrapping?

Is it the need to dynamically create `ObjectStore`s? 

I have found that dynamic creation to be one of the more complicated things about datafusion-cli (and what users of DataFusion would have to figure out to recreate it). Maybe we can figure out some simpler API for that too (as another PR)

Reply:
Maybe we can provide something like installing extensions for datafusion-cli? There're some similar use cases in DuckDB [httpfs](https://duckdb.org/docs/extensions/httpfs/overview). Before scanning the remote file, user should load the corresponding extensions (BTW, they also do some autolaoding for specific extensions, e.g. httpfs)

I think it maybe related to the `dfdb` purpose https://github.com/apache/datafusion/issues/11979 . It's a nice way to provide more feature for the user.